### ISSUE TITLE
file templates don't trigger auto rebuild

### ISSUE THREAD

**Issue by nerdoc** [tetra-framework/tetra#74]
When a component uses template_name for file templates, changing that templates doesn't trigger rebuilding tetra.

This seems due to the fact that Django (in debug mode) recompiles templates with every page refresh, so changes are seen immediately. In components, the template building is not done by a view, hence Django does not recognize it's changed...?

This has to be investigated further.

**Comment by gsxdsm** [comment]
I can help look into this!

**Comment by gsxdsm** [comment]
Still working on a way to do this via tetra, but as a workaround:


 -- In apps.py
def watch_extra_files(sender, *args, **kwargs):
    watch = sender.extra_files.add
    watch_list = glob("app/templates/**/*.html", recursive=True)
    for file in watch_list:
        if os.path.exists(file): 
            watch(Path(file))

class AppConfig(AppConfig):
    name = 'app'

    def ready(self):
        autoreload_started.connect(watch_extra_files)

**Comment by gsxdsm** [comment]
So we could put this in tetra/apps.py - maybe the templates directory could be a setting? Thoughts @nerdoc ?

**Comment by gsxdsm** [comment]
TETRA_TEMPLATES_PATH - if that is set, we setup the watcher extra files, otherwise we do nothing

**Comment by gsxdsm** [comment]
```python
from pickle import NONE
from django.apps import AppConfig
from pathlib import Path
from glob import glob
import os
from django.utils.autoreload import autoreload_started
from .templates import monkey_patch_template

monkey_patch_template()

def watch_extra_files(sender, *args, **kwargs):
    from django.conf import settings
    watch = sender.extra_files.add
    template_dir = getattr(settings, "TETRA_TEMPLATE_DIR", "app/templates")
    # List of file paths to watch
    watch_list = glob(f"{template_dir}/**/*.html", recursive=True)
    for file in watch_list:
        if os.path.exists(file):
            watch(Path(file))

class TetraConfig(AppConfig):
    name = "tetra"

    def ready(self):
        from .component_register import find_component_libraries
        from . import default_settings
        from django.conf import settings



        for name in dir(default_settings):
            if name.isupper() and not hasattr(settings, name):
                setattr(settings, name, getattr(default_settings, name))

        if not hasattr(settings, "TETRA_ESBUILD_PATH"):
            bin_name = "esbuild"
            if os.name == "nt":
                bin_name = "esbuild.cmd"
            if settings.BASE_DIR:
                setattr(
                    settings,
                    "TETRA_ESBUILD_PATH",
                    Path(settings.BASE_DIR) / "node_modules" / ".bin" / bin_name,
                )
            else:
                setattr(settings, "TETRA_ESBUILD_PATH", None)

        find_component_libraries()
        if hasattr(settings, "TETRA_TEMPLATE_DIR"):
            autoreload_started.connect(watch_extra_files)

**Comment by nerdoc** [comment]
> ```python
> def watch_extra_files(sender, *args, **kwargs):
>     watch = sender.extra_files.add
>     watch_list = glob("app/templates/**/*.html", recursive=True)
>     ...
> 
> class AppConfig(AppConfig):
>     name = 'app'
>     ...
> ```

That's funny - I didn't know about how tetra uses the autoreload API. Your code uses a good way to jump into it.
But, your  example is nailed to "app", and is accessible globally by all apps then. 

I think it is necessary that templates are NOT in a global template directory. This does not scale at all, and makes it impossible to use template in a plugin based application - where django apps bring their own templates (and components)

I definitely like the idea of using templates as files, even more than inline. Sam Willis originally made inline html templates work very cool within the BasicComponent python class, but it really has drawbacks, e.g. the caching, no translations possible in inline django_html code etc.

I for myself I like the idea very much @emilstenstrom used to compose [django_components](https://github.com/EmilStenstrom/django-components?tab=readme-ov-file#create-your-first-component), in a dir:
```
├── components/
    │   └── calendar/
    │       ├── calendar.py
    │       ├── script.js
    │       ├── style.css
    │       └── template.html  
```
So a close proximity of concerns is also reached. And a sane separation of concerns (html, js, css, py code) is done properly.
I am still inclined to implement that part for tetra too - one directory for one component. I would just change a thing, compared to django_components: the entry file should not be named *script.js*, *style.css* and *template.html* - with this you'll end up with 100 files named "template.html", all spreaded over your directories. when you are searching for the `calendar` component and search in your favourite IDE for "calendar", you  won't find the component, all template files have the same name.
So I think they should be named *calendar.html*, *calendar.js*, *calendar.css*, and *calendar.html*

Back to topic, I just wanted to give you a view about where the road maybe is going to, and you maybe can think in better terms of how to  design the code here.

#### TL;DR:
Assume that templates are at least part of a django app, and (on the <s>long</s> middle term) components could be directories with a template html file under e.g. "components". But definitely not globally. This may be added additionally (like django does), but I think it does not make too much sense, as we are a component library, right? :smile:

Maybe you could modify your code to recognize each app's template dir (this could be changed later easily when using the  "components" dir.

**Comment by EmilStenstrom** [comment]
Chiming in from django_components here: I think using django_components Component class in tetra could solve the OPs question. We have autoreload enabled for components. The changes you wanted to do to the django_component structure are all configurable. 

> I think it is necessary that templates are NOT in a global template directory. This does not scale at all, and makes it impossible to use template in a plugin based application - where django apps bring their own templates (and components)

By default we read both from the global templates directory AND from all app directories inside apps. This means that you can have both global components, and app-specific ones, and mix them together as you wish. 

We also have support for writing component libraries, where you can implement a set of components and then redistribute them: https://emilstenstrom.github.io/django-components/latest/#writing-component-libraries

> I definitely like the idea of using templates as files, even more than inline. Sam Willis originally made inline html templates work very cool within the BasicComponent python class, but it really has drawbacks, e.g. the caching, no translations possible in inline django_html code etc.

We support both templates as files and inline templates. You can mix them as you see fit.

> I am still inclined to implement that part for tetra too - one directory for one component. I would just change a thing, compared to django_components: the entry file should not be named _script.js_, _style.css_ and _template.html_ - with this you'll end up with 100 files named "template.html", all spreaded over your directories. when you are searching for the `calendar` component and search in your favourite IDE for "calendar", you won't find the component, all template files have the same name. So I think they should be named _calendar.html_, _calendar.js_, _calendar.css_, and _calendar.html_

The entry is really the calendar.py file. That file points to the names of all the components, so using calendar.css/html/js is just a search replace away. This was the original default format, and is still supported well. The reason why we switched away was to make components easier to rename, more DRY. But the other format is easily accessible.

Happy to answer any questions if you consider basing tetra on django_components.

**Comment by gsxdsm** [comment]
Thanks for the feedback - agree I strongly prefer a model where the component python, js, html, css are all co-located in the same directory.

My second code snippet can be pretty easily adapted to:

1. Reload all html (and probably js/css in the future) files located in component directories that are found via find_component_libraries
2. Also read "TETRA_TEMPLATE_DIRS" which could be an array of additional directories to watch (either a global template directory, app specific directories, or whatever the user wants).


Is this closer to what you're thinking?

**Comment by gsxdsm** [comment]
> That's funny - I didn't know about how tetra uses the autoreload API. 

To be clear, tetra *doesn't* currently use the autoreload API - that second snippet is a proposal for how we could start using it, along with a new setting I introduced (TETRA_TEMPLATE_DIRS).  Ignore https://github.com/tetra-framework/tetra/issues/74#issuecomment-2434139938 - that was just an experiment to see if it would work, not a proposal :-)

**Comment by gsxdsm** [comment]
> Chiming in from django_components here: I think using django_components Component class in tetra could solve the OPs question. We have autoreload enabled for components. The changes you wanted to do to the django_component structure are all configurable.
> 
> > I think it is necessary that templates are NOT in a global template directory. This does not scale at all, and makes it impossible to use template in a plugin based application - where django apps bring their own templates (and components)
> 
> By default we read both from the global templates directory AND from all app directories inside apps. This means that you can have both global components, and app-specific ones, and mix them together as you wish.
> 
> We also have support for writing component libraries, where you can implement a set of components and then redistribute them: https://emilstenstrom.github.io/django-components/latest/#writing-component-libraries
> 
> > I definitely like the idea of using templates as files, even more than inline. Sam Willis originally made inline html templates work very cool within the BasicComponent python class, but it really has drawbacks, e.g. the caching, no translations possible in inline django_html code etc.
> 
> We support both templates as files and inline templates. You can mix them as you see fit.
> 
> > I am still inclined to implement that part for tetra too - one directory for one component. I would just change a thing, compared to django_components: the entry file should not be named _script.js_, _style.css_ and _template.html_ - with this you'll end up with 100 files named "template.html", all spreaded over your directories. when you are searching for the `calendar` component and search in your favourite IDE for "calendar", you won't find the component, all template files have the same name. So I think they should be named _calendar.html_, _calendar.js_, _calendar.css_, and _calendar.html_
> 
> The entry is really the calendar.py file. That file points to the names of all the components, so using calendar.css/html/js is just a search replace away. This was the original default format, and is still supported well. The reason why we switched away was to make components easier to rename, more DRY. But the other format is easily accessible.
> 
> Happy to answer any questions if you consider basing tetra on django_components.

Thanks Emil! Great insight and I see your approach in django_components is not far off from what I'm proposing (we both found the same stackoverflow thread!). This discussion is also good context - https://github.com/tetra-framework/tetra/issues/74#issuecomment-2436410898

If you don't mind I plan to adjust my approach to be a bit closer to what you're doing. @nerdoc - thoughts?

**Comment by nerdoc** [comment]
> To be clear, tetra _doesn't_ currently use the autoreload API - [...] - that was just an experiment to see if it would work, not a proposal :-)

Ah, yes, technically it's just hooking into AppConfig.ready() - hooking into the autoreload API would be really cool then.

@EmilStenstrom Ah, ok, that's a much better approach with .py as entry file, and right, renaming a component should be made easy, and the less files you rename, the better. 

My thoughts about merging tetra/django_components (which is already a bit off-topic):
* I think, especially in open source projects, working together is *always* better than the currently hyped business model "My NIH-project is better than all the others". So I like the idea very much to work together as close as possible, as much of the code is somehow duplicated, at least in function. There are a few things to consider. And a few ways how to combine them:
  * Tetra could e.g. replace its BasicComponent with DC, and add additional functionality - the `Component` API on top of it. I think this approach is feasible and wouldn't even change Tetra's API. I like the "library" concept of Tetra very much, although the API is a bit cumbersome. Putting components into directories would solve the "library" approach too - Tetra's internal usage of the Library class could even be kept, the whole thing just needs refactoring then.
  * Tetra could merge with django_components, one project is abandoned. Hm, this is maybe not the best way to deal with it.

@EmilStenstrom how do you think about it? I always thought in a way that duplicating work in open source projects is a BadIdea™. It would be some effort to align Tetra's and DC's aims, but I think it would be worth it. It would also allow users to use a framework of their choice (DC for basic "stacking" of components within HTML, Tetra+DC for additional "reactiveness").

There could be other benefits, as Tetra is very opinonated with its way of holding state, there may be other on-top frameworks emerging that keep state on the server etc., all basing on DC. DC is very flexible ATM regarding configurability. I for myself like the approach better to use one path that works for 95% of the users, and works well, and leave not too much open for "cosmetic" configuration, as it basically makes codebases that use different approaches (`{% component "foo_bar" ... %}`, `{% FooBar ...%}`) at least *optically* incompatible or incoherent. But flexibility has advantages too.

But it would mean we'd have to work together more closely and help out each other maybe (which, for my part, is perfectly ok).

...my 2 cents.

**Comment by nerdoc** [comment]
OTOH, django_components already incldues some javascript parts, and has a big codebase with many possibilities, and Tetra is more streamlined with a much smaller codebase, and less option diversities.
This is not easy to combine, as @EmilStenstrom won't like to reduce features, and I wouldn't like to add too much config possibilities, but rather keep the KISS principle...

**Comment by gsxdsm** [comment]
Okay I'll create a pull request to at least solve the immediate template reloading issue. 

I like the idea of having auto discovery and components grouped by directory, but I also like the simplicity of tetra so could go either way on integration. I do feel like the goals of the project are slightly sifferent, but happy to help anywhere I can in that. Maybe a good discussion thread to pull into discussions.

**Comment by gsxdsm** [comment]
Okay I'll create a pull request to at least solve the immediate template reloading issue. I will make it read from global and app specific template locations (similar to Django components) by default but create a setting that can take an explicit list of directories to watch that will override that default. Ideally it should just work without any config. 

I like the idea of having auto discovery and components grouped by directory, but I also like the simplicity of tetra so could go either way on integration. I do feel like the goals of the project are slightly sifferent, but happy to help anywhere I can in that. Maybe a good discussion thread to pull into discussions.

**Comment by nerdoc** [comment]
I created a [discussion](https://github.com/tetra-framework/tetra/discussions/75) where the directory layout components can be planned.

**Comment by nerdoc** [comment]
@gsxdsm I am working at an alternative template loading system too - maybe we could join forces here and not duplicate the work. Could we maybe continue discussing this in #75?

**Comment by nerdoc** [comment]
I'll mark this issue as closed, as @gsxdsm's #77 closes this issue - it is still in the `components_as_dirs` branch, which will be merged ASAP. There are a few showstoppers, summarized in https://github.com/tetra-framework/tetra/issues/76.

Thank you very much.
